src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -285,11 +301,47 @@ protected PathFragment readSymbolicLink(PathFragment path) throws IOException {
   @Override
   protected void createSymbolicLink(PathFragment linkPath, PathFragment targetFragment)
       throws IOException {
-    /*
-     * TODO(buchgr): Optimize the case where we are creating a symlink to a remote output. This does
-     * add a non-trivial amount of complications though (as symlinks tend to do).
-     */
-    downloadFileIfRemote(execRoot.getRelative(targetFragment));
+    if (!linkPath.startsWith(outputBase)) {
+      throw new IOException("RemoteActionFileSystem cannot createSymbolicLink outside the output base");
+    }
+
+    Artifact output = outputMapping.get(linkPath.relativeTo(execRoot));
+    checkNotNull(output, "RemoteActionFileSystem can only create a symlink as an action output");
+
+    // If the output is a regular file or tree artifact, it has "copy" semantics, i.e., its contents
+    // are identical to the target artifact.
+    //
+    // When the target artifact is stored remotely, instead of creating a symlink in the local
+    // filesystem (which would necessarily dangle), we inject a special type of metadata that is
+    // later used by the input prefetcher to download the original artifact and put the symlink in
+    // place.
+    //
+    // Note that this implies createSymbolicLink() followed by readSymbolicLink() won't round trip,
+    // but that's ok as internal actions aren't expected to read the symlinks they create.
+    if (output.isTreeArtifact()) {
+      TreeArtifactValue metadata = getRemoteInputTreeMetadata(targetFragment);
+      if (metadata != null) {
+        checkState(targetFragment.isAbsolute());
+        for (Map.Entry<TreeFileArtifact, FileArtifactValue> entry : metadata.getChildValues().entrySet()) {
+          PathFragment childExecPath = linkPath.getRelative(entry.getKey().getParentRelativePath());
+          RemoteFileArtifactValue childMetadata = (RemoteFileArtifactValue) entry.getValue();
+          injectRemoteFile(childExecPath, childMetadata.getDigest(), childMetadata.getSize(), childMetadata.getActionId());
+        }
+        symlinkToTargetExecPath.put(linkPath, metadata.getSymlinkTargetExecPath().orElse(targetFragment.relativeTo(execRoot)));
+        return;
+      }
+    } else if (!output.isSymlink()) {
+      RemoteFileArtifactValue metadata = getRemoteInputMetadata(targetFragment);
+      if (metadata != null) {
+        checkState(targetFragment.isAbsolute());
+        injectRemoteFile(linkPath, metadata.getDigest(), metadata.getSize(), metadata.getActionId());
```

Comment:
Since we have a real fs now, can we instead just create the symlink on the in-memory fs? same for tree symlink

Reply:
Done.

The changes to `OutputService` and `ActionExecutionFunction` currently included in this PR will be submitted separately as an internal CL.